File: 트라이/leetcode211(Medium)-Design Add and Search Words Data Structure.py
# 9283ms(72.19%), 80.4MB(12.52%)

class TrieNode:
    def __init__(self):
        self.isWord = False
        self.children = collections.defaultdict(TrieNode)

class WordDictionary:
    def __init__(self):
        self.root = TrieNode()

    def addWord(self, word: str) -> None:
        node = self.root
        for char in word :
            node = node.children[char]
        node.isWord = True

    def search(self, word: str) -> bool:
        return self.match(self.root, 0, word)

    # DFS
    def match(self, node: TrieNode, index: int, word: str) -> bool:
        if index == len(word):
            return node.isWord
        
        char = word[index]
        if char == ".":
            for nextChar in node.children.keys():
                if self.match(node.children[nextChar], index + 1, word):
                    return True
            return False
        
        return char in node.children and self.match(node.children[char], index + 1, word)

File: 트라이/leetcode211(Medium)-Design Add and Search Words Data Structure.ts
// 2236ms(59.28%), 99.2MB(50.71%)

class TrieNode {
  isWord: boolean
  children: Map<string, TrieNode>

  constructor() {
    this.isWord = false
    this.children = new Map()
  }
}

class WordDictionary {
  root: TrieNode

  constructor() {
    this.root = new TrieNode()
  }

  addWord(word: string): void {
    let node = this.root
    for (const char of word) {
      if (!node.children.has(char)) {
        node.children.set(char, new TrieNode())
      }
      node = node.children.get(char) as TrieNode
    }
    node.isWord = true
  }

  search(word: string): boolean {
    return this.match(this.root, 0, word)
  }

  // DFS
  private match(node: TrieNode, index: number, word: string): boolean {
    if (index === word.length) {
      return node.isWord
    }

    const char = word[index]
    if (char === ".") {
      for (const nextChar of node.children.keys()) {
        if (
          this.match(node.children.get(nextChar) as TrieNode, index + 1, word)
        ) {
          return true
        }
      }
      return false
    }

    return (
      node.children.has(char) &&
      this.match(node.children.get(char) as TrieNode, index + 1, word)
    )
  }
}
